clusterloader2/pkg/measurement/common/kube_state_metrics_measurement.go

```diff
@@ -0,0 +1,214 @@
+	ksmServiceName               = "kube-state-metrics"
+	ksmSelfPort                  = 8081
+	ksmMetricsPort               = 8080
+)
+
+type ksmLatencyMeasurement struct {
+	ctx            context.Context
+	cancel         func()
+	isRunning      bool
+	namespace      string
+	serviceName    string
+	metricsPort    int
+	selfPort       int
+	initialLatency *measurementutil.Histogram
+	wg             sync.WaitGroup
+}
+
+func init() {
+	if err := measurement.Register(ksmLatencyName, CreateKSMLatencyMeasurement); err != nil {
+		klog.Fatalf("Cannot register %s: %v", ksmLatencyName, err)
+	}
+}
+
+// CreateKSMLatencyMeasurement creates a new Kube State
+// Metrics Measurement.
+func CreateKSMLatencyMeasurement() measurement.Measurement {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &ksmLatencyMeasurement{
+		namespace:   ksmNamespace,
+		serviceName: ksmServiceName,
+		selfPort:    ksmSelfPort,
+		metricsPort: ksmMetricsPort,
+		ctx:         ctx,
+		cancel:      cancel,
+	}
+}
+
+// Execute supports two actions:
+// - start - starts goroutine and queries /metrics every probeIntervalDefault interval,
+// it also collects initial latency metrics.
+// - gather - gathers latency metrics and creates a latency summary.
+func (m *ksmLatencyMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	client := config.ClusterFramework.GetClientSets().GetClient()
+	switch action {
+	case "start":
+		if m.isRunning {
+			klog.V(2).Infof("%s: measurement already running", m)
+			return nil, nil
+		}
+		// Start executing calls towards the kube-state-metrics /metrics endpoint
+		// every probeIntervalDefault until gather is called.
+		// probeIntervalDefault equals the scrapping interval we suggest.
+		// If we cannot get metrics for two minutes we are already going over
+		// the scrape interval so we should cancel.
+		m.startQuerying(m.ctx, client, probeIntervalDefault)
+		// Retrieve initial latency when first call is done.
+		m.initialLatency, err = m.retrieveKSMLatencyMetrics(m.ctx, client)
+		return nil, err
+	case "gather":
+		defer m.cancel()
+		return m.createKSMLatencySummary(m.ctx, client)
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (m *ksmLatencyMeasurement) stop() error {
+	if !m.isRunning {
+		return fmt.Errorf("%s: measurement was not running", m)
+	}
+	m.cancel()
+	m.wg.Wait()
+	return nil
+}
+
+// createKSMLatencyReport gathers the latency one last time and creates the summary based on the Quantile from the sub histograms.
+// Afterwards it creates the Summary Report.
+func (m *ksmLatencyMeasurement) createKSMLatencySummary(ctx context.Context, client clientset.Interface) ([]measurement.Summary, error) {
+	latestLatency, err := m.retrieveKSMLatencyMetrics(ctx, client)
+	if err != nil {
+		return nil, err
+	}
+	if err = m.stop(); err != nil {
+		klog.V(2).Info(err)
```

Comment:
Let's remove this log (the error will be logged on higher levels anyway - it's propagated out of measurement).

Reply:
Done, makes sense!